### ISSUE TITLE
merging 2 streams from file

### ISSUE THREAD

**Issue by SvenDowideit** [amsa-code/risky#4]
have hit a problem that I'm struggling to figure out.

I'm basically wanting to merge 2 file streams in an interleaved fashion - we have 2 feed files containing data from overlapping times..

I was expecting that flatMap would interleave streams from all the files, but it looks like its very much sequential :/
```
            .flatMap(file -> {
                System.err.format("FILE: %s\n", file.getName());
                return Streams.nmeaFrom(file);
            })
```
What have I missed?

**Comment by SvenDowideit** [comment]
(from @davidmoten )

Hi Sven
flatMap won't interleave streams when applied synchronously. To get interleave use multiple threads (and maxConcurrent parameter is useful to limit concurrency):

```
.flatMap(file  -> Streams.nmeaFrom(file).subscribeOn(Schedulers.io()), maxConcurrent)
```

Cheers
Dave

**Comment by SvenDowideit** [comment]
so far, I've had no success with `.subscribeOn(Schedulers.io())` - I go from a processing stream that has data, to one that doesn't. 

ie, 
```
            .flatMap(file -> {
                System.err.format("FILE: %s\n", file.getName());
                return Streams.nmeaFrom(file)
                    .subscribeOn(Schedulers.io()
                        );
            }, 2)
            .doOnNext(System.err::println)
```
only prints _2_  `FILE:` lines, and then exits. (yes, it prints as many FILE entries as I use for `maxConcurrent` - I was also expecting it to process all files.

still confused :/

**Comment by davidmoten** [comment]
Can I see more code? When you use a `subscribeOn` things happen asynchrounously.

**Comment by SvenDowideit** [comment]
my next question is - is there a way to merge the streams in the timestamp order?

(should I make a new issue for that?)

**Comment by davidmoten** [comment]
> clearly, I missed a "you need toBlocking if you're going to use subscribeOn"

In general you avoid blocking as much as possible to get reactive benefits but if you are just running a test then yep you may need to block to see what happened.

**Comment by davidmoten** [comment]
>my next question is - is there a way to merge the streams in the timestamp order?

Yep, there is. Use https://github.com/davidmoten/rxjava-extras#transformersorderedmergewith.

**Comment by SvenDowideit** [comment]
@davidmoten I'm failing again :/ how do I apply that transform to the stream above?

**Comment by davidmoten** [comment]
This is one possibility for you but I don't know where you want to end up:
```java
Observable<Timestamped<AisMessage>> a = Streams.extractMessages(Streams.fromFile(file1));
Observable<Timestamped<AisMessage>> b = Streams.extractMessages(Streams.fromFile(file2));

Comparator<Timestamped<AisMessage>> comparator =  
    (x, y) -> Long.compare(x.time(), y.time());

Observable<AisMessage> merged = a.compose(Transformers.orderedMergeWith(b, comparator));
```

**Comment by SvenDowideit** [comment]
I was hoping to avoid dropping to non-streaming style like in your example, and instead to be able to insert it into the filename->nmeastream->fanout

```
      MyApp.getAllFiles()
            .filter( reduce to the files that cover the subset we want )
            .map(file -> {
                System.err.format("FILE: %s\n", file.getName());
                return Streams.nmeaFrom(file);
            })
            .flatMap(s -> s.subscribeOn(Schedulers.io()))
            .compose(o -> Streams.extract(o))
            .filter(m -> m.getMessage().isPresent())
            .map(m -> m.getMessage().get())
            .filter(...)
... groupby, flatmap and then kinda map reduce algo
            .subscribe();
```

breaking

**Comment by davidmoten** [comment]
Things to consider
* do you want the records sorted by arrival time or record time (satellite detections can be hours old when they arrive)?
* `OrderedMerge` only works when the input streams are already sorted (so based on arrival time you can leave the input files unchanged but based on position time you would have to sort the file first and the files can get big!)
* The arrival time is the `a` field in the tag block in epoch ms, the record timestamp is the `c` field in epoch seconds
* if you want to combine more than 2 files in this way then ideally you would merge groups (of n files) then repeat till you ended up with one file. RxJava does not have an operator to repeatedly reduce a stream so to achieve that so you might use RxJava in blocking style inside a loop. I had a stab at such an operator a year ago or so and I might return to it.

**Comment by davidmoten** [comment]
Can you paste a chunk of your NMEA data here so I can see what's in it?

**Comment by davidmoten** [comment]
>I don't see any mention of a and c in the README at https://github.com/amsa-code/risky/tree/master/ais - and they're not exactly grep able - it feels like there's a huge amount of info i'm missing :/.

The project is not for documenting the NMEA and AIS standards (explaining tag blocks and the meaning of the tags) but it might be a good idea. Actually they are grepable, search for "a" and "b" with the quotes.

Yep the whole sorting business is tricky with a live feed. What is your use case for the data? You can give me a ring at AMSA (just ask switch for me) if you like.

You might want to look at the https://github.com/amsa-code/risky/tree/master/formats module as that's how I process full resolution data very fast (a year can be scanned in 20s).

**Comment by SvenDowideit** [comment]
and for anyone finding this - http://catb.org/gpsd/AIVDM.html#_nmea_tag_blocks ...

**Comment by SvenDowideit** [comment]
mmm, and after another pile of time reading the code 

the `c` tag's record time (aka `NmeaMessage.getUnixTimeMillis()` is (i think) also the `m.time()` coming in the `Streams.extract()` - and that is what we need for now.

I'm still very much interested in would re-write your `orderedMergeWith` example to be inline after/instead of

```
.flatMap(s -> s.subscribeOn(Schedulers.io()), 2)
            .compose(o -> Streams.extract(o))
```
I'm wondering if I'm basically asking how to use `Transformers.toOperator` to orderedMergeWith the current 2 observables from the flatmap - but at the same time, i'm feeling like the question shows i'm not getting "it"

**Comment by davidmoten** [comment]
>I'm still very much interested in would re-write your orderedMergeWith example to be inline 

This is why I wrote this response:

> if you want to combine more than 2 files in this way then ideally you would merge groups (of n files) then repeat till you ended up with one file. RxJava does not have an operator to repeatedly reduce a stream so to achieve that so you might use RxJava in blocking style inside a loop. I had a stab at such an operator a year ago or so and I might return to it.

To expand, suppose you have 1000 files you want to merge. You might decide to merge 5 at a time to a file then merge the 200 similarly to 40 then to 8 then to 2 then to one. This is not easily expressable in RxJava (keep transforming the Observable/Flowable and stop when we only have one item in the stream).

If you've got a smallish number of files you can merge them all in one command

```java
MyApp.getAllFiles()
     .filter( reduce to the files that cover the subset we want )
     .map(file -> Streams.nmeaFrom(file)
                 .compose(o -> Streams.extract(o))
                 .filter(m -> m.getMessage().isPresent())
                 .map(m -> m.getMessage().get())
                 .subscribeOn(Schedulers.io()))
     .toList()
    .flatMap(list -> OrderedMerge.create(list, comparator));
```

Note that `OrderedMerge` is internal api but just use it for now. The public api for ordered merge is in the `Obs` class but the methods are wrongly named (`create`!). I'll fix that in an immediate release.

**Comment by davidmoten** [comment]
rxjava-extras 0.8.0.17 has been deployed to Maven Central and will be available for download in about ten minutes (usually). Now you can replace `OrderedMerge.create` with `Obs.orderedMerge`.

**Comment by SvenDowideit** [comment]
oh! `.toList()` - though I guess that basically de-streams the data (not a big deal for a list of files) - cool!

I feel weird about doing that - I guess because I'm thinking about what would happen if I did that to a years worth of nmea data - yup, don't have the mindset for this yet :)

**Comment by SvenDowideit** [comment]
(and to answer that - nope, `toList()` will evaluate the stream to completion, and then pass that on - so I need to do it differently), lets see if i can make a functional eg

**Comment by davidmoten** [comment]
toList only evaluates the stream of *streams* to completion which is a nothing because at that stage none have been subscribed to.

**Comment by davidmoten** [comment]
The real bugbear with the approach is that it will sort in memory all of the files at once (how big are your files?).

**Comment by davidmoten** [comment]
For big files you really have to use the file system for sorting.

**Comment by SvenDowideit** [comment]
oh, newbie reading error? Can I flip your eg around to this?

(yes, I'm assuming I'm too big to sort - I presume there's an exception i can catch and deal with if the order is bad :))

```
MyApp.getAllFiles()
     .filter( reduce to the files that cover the subset we want )
     .map(file -> Streams.nmeaFrom(file)
           .compose(o -> Streams.extract(o))
           .toList()
           .flatMap(list -> OrderedMerge.create(list, comparator)))
                 .filter(m -> m.getMessage().isPresent())
                 .map(m -> m.getMessage().get())
                 // sort the records which may use a lot of memory!
                 //.sort(comparator)
                 .subscribeOn(Schedulers.io());
```

**Comment by davidmoten** [comment]
That won't work. When you do toList you end up with a List<Timestamped<AisMessage>> or similar and that is not a List of streams so OrderedMerge cannot be applied to it.

Yeah you could do the ordered merge in a flatmap but you would effectively end up with the less efficient

streams a, b, c, d

orderedMerge(d, orderedMerge(c, orderedMerge(a, b)))

There's less overhead in orderedMerge(a,b,c,d).

**Comment by SvenDowideit** [comment]
ug - I shouldn't postulate code for a language and paradigm i don't know in a github textarea - merci! - i need to play a little

**Comment by davidmoten** [comment]
The data preparation approach I took to analyse large amounts of data was to convert the nmea position messages to a fixed length binary format (custom) and append each binary record to a file particular to the vessel id (MMSI). Then I went through and sorted those individual files by timestamp. Less sorting work and the data is then easily processed in parallel.

Unfortunately you've only touched the surface with AIS. You also have to deal with getting ship static data (in a separate message emitted less frequently) and associating it with AIS position messages (so you can say at this timestamp and position the vessel had this name, IMO, draught etc).

**Comment by davidmoten** [comment]
>Yeah you could do the ordered merge in a flatmap

I think I'm wrong there, you would use `reduce` and it would have that same effect that I mentioned and be identical in behaviour to the `toList` method.
